src/platform/datapath_epoll.c

```diff
@@ -2112,127 +2136,128 @@ CxPlatSocketSendInternal(
+            CXPLAT_DBG_ASSERT(CMsg != NULL);
+            if (RemoteAddress->Ip.sa_family == QUIC_ADDRESS_FAMILY_INET) {
+                CMsg->cmsg_level = IPPROTO_IP;
+                CMsg->cmsg_type = IP_PKTINFO;
+                CMsg->cmsg_len = CMSG_LEN(sizeof(struct in_pktinfo));
+                PktInfo = (struct in_pktinfo*) CMSG_DATA(CMsg);
+                // TODO: Use Ipv4 instead of Ipv6.
+                PktInfo->ipi_ifindex = LocalAddress->Ipv6.sin6_scope_id;
+                PktInfo->ipi_addr = LocalAddress->Ipv4.sin_addr;
+            } else {
+                CMsg->cmsg_level = IPPROTO_IPV6;
+                CMsg->cmsg_type = IPV6_PKTINFO;
+                CMsg->cmsg_len = CMSG_LEN(sizeof(struct in6_pktinfo));
+                PktInfo6 = (struct in6_pktinfo*) CMSG_DATA(CMsg);
+                PktInfo6->ipi6_ifindex = LocalAddress->Ipv6.sin6_scope_id;
+                PktInfo6->ipi6_addr = LocalAddress->Ipv6.sin6_addr;
+            }
         }
     }
 
-    //
-    // Check to see if we need to pend.
-    //
-    if (!IsPendedSend) {
-        CxPlatLockAcquire(&SocketContext->PendingSendContextLock);
-        if (!CxPlatListIsEmpty(&SocketContext->PendingSendContextHead)) {
-            CxPlatSocketContextPendSend(
-                SocketContext,
-                SendData,
-                LocalAddress,
-                RemoteAddress);
-            SendPending = TRUE;
-        }
-        CxPlatLockRelease(&SocketContext->PendingSendContextLock);
-        if (SendPending) {
-            Status = QUIC_STATUS_PENDING;
-            goto Exit;
-        }
-    }
+    while (SendData->SentMessagesCount < TotalMessagesCount) {
 
-    SentByteCount = sendmsg(SocketContext->SocketFd, &Mhdr, 0);
+        int SuccessfullySentMessages =
+            sendmmsg(
+                SocketContext->SocketFd,
+                Mhdrs + SendData->SentMessagesCount,
+                (unsigned int)(TotalMessagesCount - SendData->SentMessagesCount),
+                0);
 
-    if (SentByteCount < 0) {
-        if (errno == EAGAIN || errno == EWOULDBLOCK) {
-            if (!IsPendedSend) {
-                CxPlatLockAcquire(&SocketContext->PendingSendContextLock);
-                CxPlatSocketContextPendSend(
-                    SocketContext,
-                    SendData,
-                    LocalAddress,
-                    RemoteAddress);
-                CxPlatLockRelease(&SocketContext->PendingSendContextLock);
-            }
-            SendPending = TRUE;
-            struct epoll_event SockFdEpEvt = {
-                .events = EPOLLIN | EPOLLOUT | EPOLLET,
-                .data = {
-                    .ptr = &SocketContext->EventContexts[QUIC_SOCK_EVENT_SOCKET]
+        CXPLAT_FRE_ASSERT(SuccessfullySentMessages != 0);
+
+        if (SuccessfullySentMessages < 0) {
+            if (errno == EAGAIN || errno == EWOULDBLOCK) {
+                if (!IsPendedSend) {
+                    // TODO log sends as pending so byte counts line up.
```

Comment:
I was thinking about this, and I feel we should log sends immediately as they are passed down from the caller. Abstract away if they're pending or not. Just log it and be done with it.

Reply:
We do. The issue with pending is since this path is sent again, you'll get the same log multiple times, so your sends won't line up. This todo was to add another log that a send was pended, so everything could be correlated.

I could also just make pending sends not log when going through the top of this function. That might be better.

Reply:
Yeah, that's what I was getting at.